10_rest/n0npax/cmd/puppy-server/main_test.go

```diff
@@ -0,0 +1,99 @@
+package main
+
+import (
+	"bytes"
+	"errors"
+	"fmt"
+	"os"
+	"testing"
+	"time"
+
+	"bou.ke/monkey"
+
+	puppy "github.com/anz-bank/go-course/10_rest/n0npax/pkg/puppy"
+	store "github.com/anz-bank/go-course/10_rest/n0npax/pkg/puppy/store"
+	"github.com/gin-gonic/gin"
+	"github.com/stretchr/testify/assert"
+)
+
+func TestMainWithFakeArgs(t *testing.T) {
+	fakeExit := func(int) {
+		panic("foo-arg")
+	}
+	patch := monkey.Patch(os.Exit, fakeExit)
+	defer patch.Unpatch()
+
+	os.Args = []string{"", "-s", "map", "--foo"}
+	assert.PanicsWithValuef(t, "foo-arg", main, "foo-arg")
+}
+
+func TestMain(t *testing.T) {
+	fakeRunSrv := func(*gin.Engine, int) {
+		panic("main")
+	}
+	patch := monkey.Patch(runSrv, fakeRunSrv)
+	defer patch.Unpatch()
+
+	os.Args = []string{"", "-s", "map", "-d", "../../test-data/data.json"}
+	assert.PanicsWithValuef(t, "main", main, "main")
+}
+
+func TestCreateStorer(t *testing.T) {
+	stores := []string{"map", "sync", "db"}
+	store.LevelDBPath = fmt.Sprintf("/tmp/test/%d", time.Now().UnixNano())
+	for _, v := range stores {
+		v := v
+		t.Run(v, func(t *testing.T) {
+			s := createStorer(v)
+			_, ok := s.(puppy.Storer)
+			assert.True(t, ok)
+		})
+	}
+	t.Run("dummy", func(t *testing.T) {
+		assert.PanicsWithValuef(t, "unknown storer type", func() { createStorer("dummy") },
+			"Expected Panic for unknown storer type")
+	})
+}
+
+func TestRunSrv(t *testing.T) {
+	fakeExit := func(int) {
+		panic("RunSrv")
+	}
+	patch := monkey.Patch(os.Exit, fakeExit)
+	defer patch.Unpatch()
+
+	assert.PanicsWithValuef(t, "RunSrv", func() { runSrv(nil, -22) },
+		"Expected Panic for nil in RunSrv")
+}
+
+func TestCheckError(t *testing.T) {
+	fakeExit := func(int) {}
+	patch := monkey.Patch(os.Exit, fakeExit)
+	defer patch.Unpatch()
+
+	var buf bytes.Buffer
+	out = &buf
+
+	checkError(errors.New("test"))
+	actual := buf.String()
+	assert.Equal(t, "test\n", actual)
+}
+
+func TestMainReadPuppies(t *testing.T) {
+	puppies := readPuppies("../../test-data/data.json")
+	for _, p := range puppies {
+		assert.Equal(t, 0, p.ID)
+	}
+}
+
+func TestFeedStore(t *testing.T) {
+
+	ps := readPuppies("../../test-data/data.json")
```

Comment:
nicer not to depend on path and file and execution context.
with a reader or byte slice as parameter to puppy you can avoid this.